### DISCUSSION TITLE
我在播放Demo里的"replacement.pag"文件的时候，尝试用MSAA, 但无效果

### DISCUSSION THREAD

**Discussion by hjm1fb** [Tencent/libpag#2412] [General] [answered]
你好~ 我看libpag里有MSAA的代码，所以想试用一下MSAA的效果。测试的PAG文件是"replacement.pag"。然后我做了如下的改动:

<img width="666" alt="截屏2024-08-08 18 54 50" src="https://github.com/user-attachments/assets/75e78d7f-1f95-4b1a-94dd-279a17c17366">

发现`aaType = AAType::Coverage;`的时候，画面基本是没有锯齿，但 
`aaType = AAType::MSAA`
与
`aaType = AAType::None`
的效果一样，在某些旋转角度时,即播放"replacement.pag"时设置进度为:
```
void PAGPlayer::setProgress(double percent) {
    percent = 0.442316f;
```
画面有锯齿。
然后我在其他的cpp代码里也改动代码，尽量去启用msaa的流程，发现还是有锯齿。
请问是libpag的MSAA功能还在开发阶段么？如果MSAA的功能已经Ready，我能否在播放Demo里的"replacement.pag"文件时，去掉AAType::Coverage，看看单独靠AAType::MSAA来抗锯齿的效果？

**Comment by domchen** [comment, marked as answer]
MSAA是依赖sampleCount数量大于1才能启用的，if的判断条件里就有。RT没有正确配置SampleCount当然开启不了。你改的这个变量是纯读取作用，用sampleCount数量判断当前的MSAA开启状态，好跳过CoverageAA的代码。正确的做法是改初始化RT的地方。

**Reply by hjm1fb** [reply]
看了一下，SampleCount的代码层级很多，很多地方都有usage, 都配置的话比较麻烦。有没有简单的素材配置或者代码修改，来看一下PAG的MSAA的抗锯齿效果？

**Reply by domchen** [reply]
把这个类里所有出现sampleCount的函数参数，都加一行强制赋值为4就行。https://github.com/Tencent/tgfx/blob/main/src/gpu/proxies/RenderTargetProxy.cpp
可以这么改，但没必要。对你测试的那个replacement.pag来说，只有一张图片。会出现锯齿的地方只有边缘。这个场景MSAA效果还不如现在默认的CoverageAA，边缘已经有一像素渐变过渡了。
